docs(godot): Document scopes and SentrySDK.with_scope() (#18930)
getsentry/sentry-docs: docs/platforms/godot/enriching-events/scopes/index.mdx

</Alert>

## Scopes and Threads

The current scope belongs to the thread that created it. Write to it from another thread and the SDK rejects the call with an error and drops the data.

This shapes how you use `with_scope()` around threaded work. Each thread starts with its own current scope, so a fork you create on the main thread doesn't cover a `WorkerThreadPool` task or a `Thread` you started from inside it. Call `with_scope()` inside the threaded function instead.

Errors and logs that the SDK captures automatically are enriched by the scope in effect on the thread that raised them. Data you need on every thread belongs on the global scope.

## Scopes in C#

The .NET layer keeps its own scope model, driven by the Sentry .NET SDK API and available since C# support landed in 2.0.0. Call it through `Sentry.Godot.SentrySdk`, the same way you'd call any other .NET SDK method:

```csharp
SentrySdk.ConfigureScope(scope =>
{
    scope.SetTag("subsystem", "savegame");
});

// Only this message is tagged with the slot.
SentrySdk.CaptureMessage("Save failed", scope => scope.SetTag("save_slot", "3"));
```

> `ConfigureScope()` and the scope callback available on every capture method behave as documented for .NET — see [Scopes and Hubs](/platforms/dotnet/enriching-events/scopes/) for the details.

Author: Nope: https://docs.sentry.io/platforms/dotnet/enriching-events/scopes/

---
title: Scopes

> description: "Learn what a scope is and how to use one to enrich only the telemetry captured by a single part of your game."

Reviewer: ```suggestion
description: "Learn what a scope is and how to use it to enrich telemetry from a specific part of your game."
```

---

<AvailableSince version="2.2.0" />

When the SDK captures an event, it merges the event data with the extra information held on the scopes that are in effect. Most of what you set through `SentrySDK` applies to your whole game, which is usually what you want. Scopes are how you narrow that down, so the data you attach while loading one mod doesn't ride along with everything your game captures for the rest of the session.

## What's a Scope?

> A scope holds the data that rides along with your telemetry: <PlatformLink to="/enriching-events/tags/">tags</PlatformLink>, <PlatformLink to="/enriching-events/context/">contexts</PlatformLink>, <PlatformLink to="/enriching-events/breadcrumbs/">breadcrumbs</PlatformLink>, <PlatformLink to="/enriching-events/identify-user/">user</PlatformLink>, severity level, fingerprint, and attributes attached to <PlatformLink to="/logs/">logs</PlatformLink> and <PlatformLink to="/metrics/">metrics</PlatformLink>. `SentryScope` provides setters for these values, except breadcrumbs, which you append with `add_breadcrumb()`. Calling `clear()` empties the scope, including any data it inherited when it was forked. Data set globally is unaffected.

Author: Both exist, `scope.add_breadcrumb` adds crumbs on the local scope, while the top-level api operates on global.

> A scope holds the data that rides along with your telemetry: <PlatformLink to="/enriching-events/tags/">tags</PlatformLink>, <PlatformLink to="/enriching-events/context/">contexts</PlatformLink>, <PlatformLink to="/enriching-events/breadcrumbs/">breadcrumbs</PlatformLink>, the <PlatformLink to="/enriching-events/identify-user/">user</PlatformLink>, the severity level, the fingerprint, and attributes attached to <PlatformLink to="/logs/">logs</PlatformLink> and <PlatformLink to="/metrics/">metrics</PlatformLink>. `SentryScope` provides setters for these values, except breadcrumbs, which you append with `add_breadcrumb()`. Calling `clear()` empties the scope, including any data it inherited when it was forked. Data set globally is unaffected.

Reviewer 2: I'm not sure, but the articles make it feel a little uneven

```suggestion
A scope holds the data that rides along with your telemetry: <PlatformLink to="/enriching-events/tags/">tags</PlatformLink>, <PlatformLink to="/enriching-events/context/">contexts</PlatformLink>, <PlatformLink to="/enriching-events/breadcrumbs/">breadcrumbs</PlatformLink>, <PlatformLink to="/enriching-events/identify-user/">user</PlatformLink>, severity level, fingerprint, and attributes attached to <PlatformLink to="/logs/">logs</PlatformLink> and <PlatformLink to="/metrics/">metrics</PlatformLink>. `SentryScope` provides setters for these values, except breadcrumbs, which you append with `add_breadcrumb()`. Calling `clear()` empties the scope, including any data it inherited when it was forked. Data set globally is unaffected.
```

</Alert>

## Using `with_scope()`

`SentrySDK.with_scope()` forks the current scope, hands the fork to your callable, and discards it when the callable returns. Everything captured inside the callable picks up the scope data, and everything after it doesn't:

```gdscript
func load_mod(mod_id: String, version: String, path: String) -> void:
	SentrySDK.with_scope(func(scope: SentryScope) -> void:
		scope.set_tag("subsys", "mods")
		scope.set_context("mod", {
			"id": mod_id,
			"version": version,
			"path": path,
		})
		_apply_mod(path)
	)
```

Mods fail to load, and when one does you need to know which. If `_apply_mod()` pushes an error, the issue arrives naming the mod and its version, so you can tell a broken mod from a bug in your loader. The next mod loads under its own scope, and once loading is done nothing carries either.

`with_scope()` returns whatever the callable returns, so you can wrap an existing call without restructuring it:

```gdscript
func save_game(slot: int) -> bool:
	var saved: bool = SentrySDK.with_scope(func(scope: SentryScope) -> bool:
		scope.set_context("save_file", {"slot": slot, "path": _save_path(slot)})
		return _write_save_file(slot)
	)

> 	return saved

Author: The actual implementation forwards the return value, backed by tests. Quite a useful feature actually.

```

> Nesting works the same way: an inner `with_scope()` forks the enclosing scope, starts with its data, and keeps its own writes to itself. Data you write through `SentrySDK` still goes to the global scope, even from inside the callable.

Reviewer: ```suggestion
Nesting works the same way: an inner `with_scope()` forks the enclosing scope, inherits its data, and keeps any changes local. Data you write through `SentrySDK` still goes to the global scope, even from inside the callable.
```

<Alert level="warning">

The fork covers only the synchronous part of the callable, including the work reached through the functions it calls. Anything that resumes after an `await` runs outside it, and you won't always get a warning. Await outside the block, then wrap the synchronous work that follows.

</Alert>

## Reading the Current Scope

> `SentrySDK.get_current_scope()` returns the scope active on the calling thread. Inside a `with_scope()` callable, that's the fork it handed you — useful when the code that has something to add isn't the code holding the `scope` parameter:

Reviewer: ```suggestion
`SentrySDK.get_current_scope()` returns the scope active on the calling thread. Inside a `with_scope()` callable, that's the fork it handed you — useful when the code adding data doesn't have access to the `scope` parameter:
```
